### ISSUE TITLE
SQS stream stops consuming messages

### ISSUE THREAD

**Issue by gipeshka** [akka/alpakka#1665]
### Versions used 

Alpakka version: 1.0.0
Akka version: 2.5.17

### Expected Behavior

Processing should go on

### Actual Behavior

What we see after update from M1 to 1.0.0 is that queue consumption stops randomly with no error reported to a supervisor.

To track errors we use a supervision policy in materializer:
```scala
  implicit val materializer: Materializer = ActorMaterializer(
    ActorMaterializerSettings(as).withSupervisionStrategy { error =>
      logger.error("Stream processing has failed", error)

      error match {
        case NonFatal(_) => Supervision.Resume
        case _ => Supervision.Stop
      }
    }
  )
```

The stream consists of a source spawned via `SqsSource.apply`, a grouping ack-flow created via `SqsAckFlow.grouped` and the business logic in between. We use the default `SqsAsyncClient` as spawned via `SqsAsyncClient.create()`. The resulting sink is simply `Sink.ignore`

What we see in thread pool monitoring is the following drop at the moment when consumption stops:

![Screen Shot 2019-04-25 at 16 15 06](https://user-images.githubusercontent.com/1704933/56742595-57992680-6775-11e9-9446-385a96260995.png)

Unfortunately I'm unable to reproduce it locally, failure only happens under high load on production servers, but it usually happens every hour or so.

**Comment by ennru** [comment]
Thank you for reporting. 
I guess we need to revisit the changes between M1 and 1.0.0.

**Comment by ranchoiver** [comment]
I've also had this problem (my comment from another issue: https://github.com/akka/alpakka/issues/1588#issuecomment-475584379)

**Comment by ennru** [comment]
For 1.0-M3 the move to AWS SDK 2.0 in #1450 replaced the source stage with the much simpler implementation we have today: https://github.com/akka/alpakka/blob/docs/v1.0-M2/sqs/src/main/scala/akka/stream/alpakka/sqs/impl/SqsSourceStage.scala

@gipeshka Just to be sure, you DON'T use `closeOnEmptyReceive`, right?

**Comment by gipeshka** [comment]
@ennru yes, we don't `closeOnEmptyReceive`.

Also according to SQS monitoring dashboard we barely have any empty receives at the time when the consumptions stops.

**Comment by Vict0r7** [comment]
I'm experiencing the exact same issue since 1.0.0 too 😕

**Comment by avdv** [comment]
Hi.

We're facing the same problem. We also see the number of threads dropping, starting at the time when no SQS messages can be received:
![image](https://user-images.githubusercontent.com/3471749/56959231-0c07c380-6b4d-11e9-8bcc-7b9a4715a78e.png)
When we stop the stream via a kill switch and switch it on again (at around 8:30), it starts to pull messages again.

**Comment by RustedBones** [comment]
If the `SqsSource` is back-pressured, no `ReceiveMessageRequest` will be sent, so the `SqsAsyncClient` wont create thread to poll the queue.
On restart of the application, the `SqsSource` will have to fill the buffer, that may explain why it works for a bit and stops afterwards.

@Vict0r7 & @avdv are you also using `SqsAckFlow` or `SqsAckSink` in batch mode ?

**Comment by Vict0r7** [comment]
@RustedBones yes, i'm using `SqsAckFlow.grouped`

**Comment by avdv** [comment]
We are using `SqsAckSink`.

BTW, we are explicitly using the `AkkaHttpAsyncHttpService` from https://github.com/matsluni/aws-spi-akka-http as the HttpClient.

We already upgraded the AWS sqs library to version 2.5.24.

**Comment by gabfssilva** [comment]
I had this error since the previous versions. When an error occurr while the client is fetching the messages, the Stream is completed with an error.

What I ended up with this solution:

```scala
def sqsSource: Source[Message, NotUsed] = _

def restartableSqsSource = 
  RestartSource
    .onFailuresWithBackoff(1 second, 5 seconds, 0.1)(() => sqsSource)
```

I guess we could make SqsSource restartable by default, since network issues can happen and most users want infinite element consuming.

**Comment by avdv** [comment]
> I had this error since the previous versions. When an error occurr while the client is fetching the messages, the Stream is completed with an error.

Did you really see an error? Because, we do not, the stream is still alive, AFAICT.  

```
stream.run() onComplete {
      case Success(_) ⇒ logger.debug(loginfo << Status.Ok)
      case Failure(e) ⇒
        logger.error(
          loginfo << Status.Failed << e << message("processing SQS messages"),
          e
        )
}
```
I would expect a message in any case when the stream is completed. But I don't see it, instead, after hitting the kill switch, I see the "OK" message.

I'll try your solution next week, maybe it makes a difference.

**Comment by avdv** [comment]
BTW, we had a short network outage today, and indeed all SQS processes stopped consuming messages for every environment. So it is related to some network / connection issue probably.

Also, we're not hitting the problem very often. Maybe once every three days.

**Comment by RustedBones** [comment]
@avdv I would be really careful using the `AkkaHttpAsyncHttpService`.
It's clearly stated to be **not production ready** in its README.

I've had a quick look at the code, and I may have found the problem:

In their [RunnableRequest](https://github.com/matsluni/aws-spi-akka-http/blob/0fae9a51093768e0d8aabec11d2c0ad624a3b1b4/src/main/scala/com/github/matsluni/akkahttpspi/RunnableRequest.scala#L37), no `handler.onError` is set. That means if the internal AWS SDK has an error, the promise will never be completed, leading to alpakka back pressuring

**Comment by RustedBones** [comment]
PR oppened for `aws-spi-akka-http`: See [#6](https://github.com/matsluni/aws-spi-akka-http/pull/6/files)

**Comment by RustedBones** [comment]
@gabfssilva I'm not sure making the `SqsSource` restartable by default is a good solution.
The restart and supervision must be kept explicit to avoid confusion IMHO.

**Comment by gabfssilva** [comment]
@RustedBones good catch. While your PR is not merged, it's probably best to use the default Netty connector instead.
I also agree about the restartable `SqsSource`, but, I guess we should find a way to explicitly say somewhere inside the docs that all Source connectors are liable to error and eventually will stop producing messages due its completion.

**Comment by avdv** [comment]
> @avdv I would be really careful using the `AkkaHttpAsyncHttpService`.
> It's clearly stated to be **not production ready** in its README.

Yes, I know. But it's actually the default with alpakka-sqs, so I figured it would be OK.

https://github.com/akka/alpakka/blob/52f6f6dd198109f468b9765a173ef638b2217e7a/project/Dependencies.scala#L374-L387

Honestly, I was also looking into the SPI as the likely culprit, but couldn't figure it out. Nice find @RustedBones :+1: !

**Comment by matsluni** [comment]
I just released version `0.0.6` with the fix from @RustedBones. The artifacts should be available soon on Maven Central.

**Comment by ennru** [comment]
Just to let you know, Alpakka SQS doesn't use `aws-spi-akka-http` by default. See #1693 

But that doesn't seem to be the problem here either, did anyone investigate this further?

**Comment by ennru** [comment]
When I look at `SqsSource` it appears to me that the `mapAsync` fires off a couple of requests simultaneously. 

https://github.com/akka/alpakka/blob/v1.0.0/sqs/src/main/scala/akka/stream/alpakka/sqs/scaladsl/SqsSource.scala#L45

Can you try and configure `settings.maxBufferSize / settings.maxBatchSize` to be 1, please. I believe it should always be 1 instead of connected to the buffer.

**Comment by RustedBones** [comment]
@ennru I'm also ok to fix the `mapAsync`to 1. There is no real advantage of setting higher parallelism since it is connected to a buffer as you stated. As previously explained [here](https://github.com/akka/alpakka/issues/1588#issuecomment-476105607), this mainly helps at startup to fill the buffer faster

**Comment by ennru** [comment]
Has anybody tried this with the fixed version of `aws-spi-akka-http`?
We plan to release Alpakka 1.0.1 on Thursday.

**Comment by avdv** [comment]
> We actually use aws-spi-akka-http by default as it is injected via reflection.

Yeah, since the netty based service implementation was explicitly excluded you only have a single service implementation on the class path by default:
```
> java.util.ServiceLoader.load(classOf[software.amazon.awssdk.http.async.SdkAsyncHttpService]).iterator.asScala.toList
res4: List[software.amazon.awssdk.http.async.SdkAsyncHttpService] = List(com.github.matsluni.akkahttpspi.AkkaHttpAsyncHttpService@44cd40d2)
```
> Has anybody tried this with the fixed version of `aws-spi-akka-http`?

Yes, we switched to 0.0.6 right after it was released, deployed to production last Wednesday. We have not seen this problem since.

We could reproduce the original problem of the SQS stream coming to a halt without error message / completion by blocking network traffic to the Amazon server(s). With the updated dependency (1) we are seeing the failed HTTP requests being logged, and (2) the stream is resumed as soon as we re-enable the network communication.

**Comment by ennru** [comment]
Great, thank you for reporting back.
I consider this closed now.

**Comment by zallesov** [comment]
I came here with the same issue. But it turned to be we just did not notice the `SQS:DeleteMessage` permission were missing on Aws side.

**Comment by shaileshntikhe** [comment]
Is this fixed? I am using 2.0.2 and observing this. Also I see drop in thread count once message consumption is stopped as @avdv [mentioned](https://github.com/akka/alpakka/issues/1665#issuecomment-487919225).
